src/Microsoft.Azure.WebJobs.Host/Executors/JobHostContextFactory.cs

```diff
@@ -27,121 +27,162 @@
 
 namespace Microsoft.Azure.WebJobs.Host.Executors
 {
-    internal class JobHostContextFactory : IJobHostContextFactory
+    internal static class JobHostConfigurationExtensions
     {
-        private readonly IStorageAccountProvider _storageAccountProvider;
-        private readonly IConsoleProvider _consoleProvider;
-        private readonly JobHostConfiguration _config;
-
-        public JobHostContextFactory(IStorageAccountProvider storageAccountProvider, IConsoleProvider consoleProvider, JobHostConfiguration config)
+        // Do full initialization (both static and runtime). 
+        // This can be called multiple times on a config. 
+        public static async Task<JobHostContext> CreateAndLogHostStartedAsync(
+            this JobHostConfiguration config, 
+            JobHost host, 
+            CancellationToken shutdownToken, 
+            CancellationToken cancellationToken)
         {
-            _storageAccountProvider = storageAccountProvider;
-            _consoleProvider = consoleProvider;
-            _config = config;
+            var newServices = config.DoStaticInitialization();
+            JobHostContext context = await newServices.DoRuntimeInitialization(config, host, shutdownToken, cancellationToken);
+
+            return context;
         }
 
-        public async Task<JobHostContext> CreateAndLogHostStartedAsync(JobHost host, CancellationToken shutdownToken, CancellationToken cancellationToken)
+        // Static initialization. Returns a service provider with some new services initialized. 
+        // The new services:
+        // - can retrieve static config like binders and converters; but the listeners haven't yet started.
+        // - can be flowed into the runtime initialization to get a JobHost spun up and running.
+        // This shouldn't be async beccause we shouldn't need to do network calls for static init.        
+        // This can be called multiple times on a config, which is why it returns a new ServiceProviderWrapper
```

Comment:
How is this called multiple times on a config? It is called only once for the lifetime of a JobHost, right? Are you talking about scenarios like Functions, where we start/stop hosts using the same config?